### PR TITLE
[Spark] Improve missing stats column message for unsupported data skipping types

### PR DESCRIPTION
**Pull request by dabao521**
#### Which Delta project/connector is this regarding?

- [x] Spark
- [ ] Standalone
- [ ] Flink
- [ ] Kernel
- [ ] Other (fill in here)

## Description

Today when either the clustering column data type is unsupported for data skipping or clustering column is not in the first 32 columns, it shows the following message

[DELTA_CLUSTERING_COLUMN_MISSING_STATS] Liquid clustering requires clustering columns to have stats. Couldn't find clustering column(s) 'current_version' in stats schema:

This is confusing to users when the column data type is not supported for data skipping. To improve this scenario this PR introduces a new error class DELTA_CLUSTERING_COLUMNS_NOT_SUPPORTED_DATATYPE  when cluster by non data skipping data type such as cluster by Boolean column

## How was this patch tested?

Existing unit tests.

### CODE REVIEW

File: spark/src/main/resources/error/delta-error-classes.json [hunk 1]
```diff
@@ -387,6 +387,12 @@
     ],
     "sqlState" : "42P10"
   },
+  "DELTA_CLUSTERING_COLUMNS_NOT_SUPPORTED_DATATYPE" : {
```
[the remark this answers is not in the document]
Comment (dabao521) [reply spark/src/main/resources/error/delta-error-classes.json]: Makes sense and I reworded it to `DELTA_CLUSTERING_COLUMNS_DATATYPE_NOT_SUPPORTED`

File: spark/src/main/scala/org/apache/spark/sql/delta/skipping/clustering/ClusteredTableUtils.scala [hunk 2]
```diff
@@ -355,11 +355,29 @@ trait ClusteredTableUtilsBase extends DeltaLogging {
...
+      getClusteringColumnsNotInStatsSchema(statsCollection, clusteringColumnInfos)
+    if (missingColumns.nonEmpty) {
+      // Check DataType eligibility.
+      val missingColumnSet = missingColumns.toSet
+      val missingColumnInfos = clusteringColumnInfos.filter(
+        info => missingColumnSet.contains(info.logicalName))
+      // This assertion must hold since missingColumns are subset of clusteringColumnInfos.
+      assert(missingColumnInfos.length == missingColumns.length)
+      val (skippingEligibleMissingColumnInfos, nonSkippingEligibleMissingColumnInfos) =
+        missingColumnInfos.partition(info => SkippingEligibleDataType(info.dataType))
```
[the remark this answers is not in the document]
Comment (dabao521) [reply spark/src/main/scala/org/apache/spark/sql/delta/skipping/clustering/ClusteredTableUtils.scala]: Good point and I reworked as suggested

File: spark/src/main/scala/org/apache/spark/sql/delta/skipping/clustering/ClusteredTableUtils.scala [hunk 3]
```diff
@@ -355,11 +355,29 @@ trait ClusteredTableUtilsBase extends DeltaLogging {
   private def validateClusteringColumnsInStatsSchema(
       statsCollection: StatisticsCollection,
       clusteringColumnInfos: Seq[ClusteringColumnInfo]): Unit = {
-    val missingColumn = getClusteringColumnsNotInStatsSchema(statsCollection, clusteringColumnInfos)
-    if (missingColumn.nonEmpty) {
+    val missingColumns =
+      getClusteringColumnsNotInStatsSchema(statsCollection, clusteringColumnInfos)
+    if (missingColumns.nonEmpty) {
+      // Check DataType eligibility.
+      val missingColumnSet = missingColumns.toSet
```
Comment (zedtang) [inline spark/src/main/scala/org/apache/spark/sql/delta/skipping/clustering/ClusteredTableUtils.scala:362]: nit: doesn't look like set is needed as Seq also supports contains. Also there'll be at most 4 clustering columns

**Review by zedtang — APPROVED** [review]

File: spark/src/main/scala/org/apache/spark/sql/delta/skipping/clustering/ClusteredTableUtils.scala (hunk 3 above)
Comment (dabao521) [reply spark/src/main/scala/org/apache/spark/sql/delta/skipping/clustering/ClusteredTableUtils.scala]: fixed

**Review by chirag-s-db — APPROVED** [review]

**Pull request merged**